### ISSUE TITLE
torch.nn.functional.cosine_similarity outputs NaN

### ISSUE THREAD

**Issue by DNXie** [pytorch/pytorch#51912]
## 🐛 Bug

`torch.nn.functional.cosine_similarity` outputs NaN

## To Reproduce

~~~Python
import torch
torch.nn.functional.cosine_similarity(x1=torch.tensor([2.0775e+38, 3.0262e+38]), x2=torch.tensor([1,1]), dim=0)
~~~

Output:
~~~python
tensor(nan)
~~~

## Environment

PyTorch version: 1.8.0.dev20210126+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.4 LTS (x86_64)
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.7 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

cc @albanD @mruberry @jbschlosser

**Comment by mruberry** [comment]
This seems expected to me. I think both the numerator and denominator of the expression aren't representable in float32, which is what you're computing in. Passing double tensors will produce the correct value.

@jbschlosser I think we close this as "by design"?

**Comment by ngimel** [comment]
I'm not sure that's expected, there are ways of computing cosine similarity that would avoid overflow, e.g. #31378 that we should really unblock and land.

**Comment by Lezcano** [comment]
I think that merging #31378 would be great, as it is implements a better approach than the one we currently have.

Now, I'm afraid that this new approach won't fix the example in this issue, as we have that the norm of `torch.tensor([2.0775e+38, 3.0262e+38]).norm()` is not representable in 32 signed bits. In my opinion, it's safe to close this issue with a recommendation to use double precision if one is going to work with these large numbers, as @mruberry already did
